core/peerset/src/lib.rs

```diff
@@ -147,156 +130,176 @@ pub struct PeersetConfig {
-				self.data.discovered.add_peer(peer_id.clone(), SlotType::Common);
-				self.message_queue.push_back(Message::Drop(peer_id));
+		self.reserved_only = reserved_only;
+		if self.reserved_only {
+			for peer_id in self.data.connected_peers().cloned().collect::<Vec<_>>().into_iter() {
+				let peer = self.data.peer(&peer_id).into_connected()
+					.expect("We are enumerating connected peers, therefore the peer is connected; qed");
+				if !peer.is_reserved() {
+					peer.disconnect();
+					self.message_queue.push_back(Message::Drop(peer_id));
+				}
 			}
+
 		} else {
 			self.alloc_slots();
 		}
 	}
 
 	fn on_report_peer(&mut self, peer_id: PeerId, score_diff: i32) {
-		let score = match self.data.scores.get_mut(&peer_id) {
-			Some(score) => {
-				*score = score.saturating_add(score_diff);
-				*score
-			},
-			None => {
-				self.data.scores.insert(peer_id.clone(), score_diff);
-				score_diff
-			}
+		// We want reputations to be up-to-date before adjusting them.
+		self.update_time();
+
+		match self.data.peer(&peer_id) {
+			peersstate::Peer::Connected(mut peer) => peer.add_reputation(score_diff),
+			peersstate::Peer::NotConnected(mut peer) => peer.add_reputation(score_diff),
+			peersstate::Peer::Unknown(peer) => peer.discover().add_reputation(score_diff),
+		}
+	}
+
+	/// Updates the value of `self.latest_time_update` and performs all the updates that happen
+	/// over time, such as reputation increases for staying connected.
+	fn update_time(&mut self) {
+		// We basically do `(now - self.latest_update).as_secs()`, except that by the way we do it
+		// we know that we're not going to miss seconds because of rounding to integers.
+		let secs_diff = {
+			let now = Instant::now();
+			let elapsed_latest = self.latest_time_update - self.created;
+			let elapsed_now = now - self.created;
+			self.latest_time_update = now;
+			elapsed_now.as_secs() - elapsed_latest.as_secs()
 		};
 
-		if score < 0 {
-			// peer will be removed from `in_slots` or `out_slots` in `on_dropped` method
-			if self.data.in_slots.contains(&peer_id) || self.data.out_slots.contains(&peer_id) {
-				self.data.in_slots.remove_peer(&peer_id);
-				self.data.out_slots.remove_peer(&peer_id);
-				self.message_queue.push_back(Message::Drop(peer_id));
+		// For each elapsed second, move the node reputation towards zero.
+		// If we multiply each second the reputation by `k` (where `k` is between 0 and 1), it
+		// takes `ln(0.5) / ln(k)` seconds to reduce the reputation by half. Use this formula to
+		// empirically determine a value of `k` that looks correct.
+		for _ in 0..secs_diff {
+			for peer in self.data.peers().cloned().collect::<Vec<_>>() {
+				// We use `k = 0.98`, so we divide by `50`. With that value, it takes 34.3 seconds
+				// to reduce the reputation by half.
+				match self.data.peer(&peer) {
+					peersstate::Peer::Connected(mut peer) => {
+						let cur_reput = peer.reputation();
+						peer.set_reputation(cur_reput.saturating_sub(cur_reput / 50));
```

Comment:
This seems to be increasing negative reputations. Also, I'm not sure we really need the `secs_diff` loop.

Reply:
> This seems to be increasing negative reputations.

That's on purpose!

Reply:
Oh, I see. But then is it possible for someone to misbehave, wait a few minutes and do it again? and what if this someone has many nodes doing it concurrently?

Reply:
> But then is it possible for someone to misbehave, wait a few minutes and do it again?

Yes, but one can also generate a brand new key at any time, so decreasing the reputation of a node cannot be a protection mechanism.

Reply:
I guess I'm having trouble understanding why it should be an exception to the principle we are following (increasing reputation only when peer provides useful information).

Reply:
If a node's reputation goes down, we will disconnect from it to replace it with a better one (that's not implemented yet). If that happens, there's no way the reputation can go back to 0 other than through this mechanism.

Reply:
How does it work the mechanism after they are taken to `0`? I still feel that `0` should't be important.

Reply:
Well, there must be a default reputation for unknown nodes. Zero is a reasonable choice for that. In my mind the only question is whether to allow a reputation to go below zero. But I think it's actually fine; it means "choose this guy only if there are no newcomers to try out".

Reply:
It's ok. I was just thinking it more like: 
- The default reputation is the mean between nodes "capable of producing something bad or good". 
- Without negative reputation (or doesn't matters). 
- Choosing to connect just by ranking.
- Without helping negative ones towards zero (that doesn't look like the mean to me).
- Without not helping (and still penalizing with time) positive nodes close to zero.

But probably my idea of it is wrong or I'm missing something.

Reply:
"the mean between nodes"

the mean between *which* nodes? all nodes we've ever seen? at which point? right now? even if that made sense (which i don't think it does), then that would suggest there be no difference between a newcomer (which we have precisely no evidence is useful and which could be trivially sybil attacked) and a pre-existing node that has behaved "ok" for a long time.

Reply:
Hmm, it could be the mean between recent N nodes. I guess in a sybil attack the mean will go down fast. I didn't think much about it.

The point is that it doesn't matter, we can use zero as default, and we can always add more information and customize the selection rule to distinguish between newcomers and stable nodes (connected timestamp?). 

On the other side, by moving negative reputation towards zero, I think we are losing information (newcomers and bad guys are equal after a X minutes) and coupling the estimation with the selection, losing flexibility. 

For instance: changing the default seems to be more than setting a variable, the rate for "forgiveness" and "time punishment" is the same, deactivating "forgiveness", etc.

Anyway, maybe they are just implementation details.